### PR TITLE
add the coin example

### PR DESCRIPTION
**Pull request by konnov**
I have finally found an example for the next tutorial. It is simple enough and yet reasonable. We can also demonstrate various features with this example.

This example took me 1:30h to do. With a bit of experience and a right set of idioms this may become quite easy to do.

- [x] Tests added for any new code
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality

### CODE REVIEW

File: examples/solidity/Coin/Coin.qnt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        // keep the minter unchanged
+        minter' = minter,
+    }
+
+    // All possible behaviors of the protocol in one action.
+    // Note that random simulation produces valid inputs to
+    // this action with a low probability, since amount are chosen at random.
+    // A symbolic model checker such as Apalache does not have any problem
+    // with this. Ranom simulation requires a more restricted version of step.
+    action step(): bool = {
```
Comment (thpani) [inline examples/solidity/Coin/Coin.qnt:117]: I guess we don't need the `()`? To be consistent with `init`:

```suggestion
    action step: bool = {
```

**Review by thpani — APPROVED** [review]
LGTM, mostly language suggestions and 2 questions on comments that are unclear to me.

NB: I haven't read the full Solidity example.

File: examples/solidity/Coin/Coin.qnt [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    // the maximal value Solidity UInt can carry
+    pure val MAX_UINT = 2^256 - 1
+
+    // does a big integer represent Solidity uint?
+    pure def isUInt(i: int): bool = 0 <= i and i <= MAX_UINT
+
+    // Currently, we fix the set of all addresses to a small set.
+    // In the future, we will be using the commented-out constant declaration.
+    // For now, we are using simple aliases instead of actual Ethereum addresses.
+    // The names that in the set CONTRACT_ADDR belong to some contracts.
```
[the remark this answers is not in the document]
Comment (konnov) [reply examples/solidity/Coin/Coin.qnt:40]: That's copy&paste from another example. Will fix.

File: examples/solidity/Coin/Coin.qnt [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    // the minter's address
+    var minter: Addr
+    // the balances in the subcurrency tokens
+    var balances: Addr -> UInt
+
+    // a helper function to make preconditions similar to Solidity
+    def require(cond: bool): bool = cond
+
+    // a handy definition to query the whole state in REPL at once
+    def state = { minter: minter, balances: balances }
```
Comment (shonfeder) [inline examples/solidity/Coin/Coin.qnt:56]: Moving the related code closer together:

```suggestion
    // a handy definition to query the whole state in REPL at once
    def state = { minter: minter, balances: balances }
    
    // a helper function to make preconditions similar to Solidity
    def require(cond: bool): bool = cond
```

File: examples/solidity/Coin/Coin.qnt [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    // All possible behaviors of the protocol in one action.
+    // Note that random simulation produces valid inputs to
+    // this action with a low probability, since amount are chosen at random.
+    // A symbolic model checker such as Apalache does not have any problem
+    // with this. Random simulation requires a more restricted version of step.
+    action step: bool = {
+        nondet sender = ADDR.oneOf()
+        nondet receiver = ADDR.oneOf()
```
Comment (shonfeder) [inline examples/solidity/Coin/Coin.qnt:119]: reads more naturally with prefix, imo

```suggestion
        nondet sender = oneOf(ADDR)
        nondet receiver = oneof(ADDR)
```

File: examples/solidity/Coin/Coin.qnt [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        initFixed
+            .then(mint("alice", "bob", 10))
+            .then(send("bob", "eve", 4))
+            .then(all {
+                assert(balances.get("bob") == 6),
+                assert(balances.get("eve") == 4),
+                minter' = minter,
+                balances' = balances,
+            })
+    }
```
Comment (shonfeder) [inline examples/solidity/Coin/Coin.qnt:158]: This is cool.

**Review by shonfeder — APPROVED** [review]
lookin good1

File: examples/solidity/Coin/Coin.qnt [hunk 6]
```diff
@@ -0,0 +1,185 @@
+/**
+ * Coin contract from the Solidity examples [1]. It is a nice starting
+ * example for Solidity. It is a nice starting example for Quint too.
+ *
+ * This contract introduces a custom token. Our protocol specification is
+ * not designed to be callable from other protocols. It is possible to make it
+ * composable, but it would introduce additional code (following simple idioms),
```
[the remark this answers is not in the document]
Comment (konnov) [reply examples/solidity/Coin/Coin.qnt:7]: Solidity contracts can be usually called from other contracts. We can achieve a similar effect in Quint, but that requires more effort. I will fix the wording there.

**Pull request merged**